relay-cabi/src/processing.rs

```diff
@@ -321,3 +325,72 @@ pub unsafe extern "C" fn relay_validate_project_config(
         Err(e) => RelayStr::from_string(e.to_string()),
     }
 }
+
+#[derive(Debug, Deserialize)]
+struct EphemeralEvent {
+    transaction: String,
+}
+
+impl EphemeralEvent {
+    fn to_event(&self) -> Event {
+        Event {
+            transaction: Annotated::new(self.transaction.clone()),
+            ty: Annotated::new(EventType::Transaction),
+            ..Default::default()
+        }
+    }
+}
+
+#[derive(Debug, Serialize)]
+struct EphemeralSamplingResult {
+    merged_sampling_configs: Vec<SamplingRule>,
+    sampling_match: Option<SamplingMatch>,
+}
+
+/// Runs dynamic sampling given the sampling config, root sampling config, dsc and event.
+///
+/// Returns the sampling decision containing the sample_rate and the list of matched rule ids.
+#[no_mangle]
+#[relay_ffi::catch_unwind]
+pub unsafe extern "C" fn run_dynamic_sampling(
+    sampling_config: *const RelayStr,
+    root_sampling_config: *const RelayStr,
+    dsc: *const RelayStr,
+    event: *const RelayStr,
+) -> RelayStr {
+    let sampling_config = serde_json::from_str::<SamplingConfig>((*sampling_config).as_str())?;
+    let root_sampling_config =
+        serde_json::from_str::<SamplingConfig>((*root_sampling_config).as_str())?;
+    // We can optionally accept a dsc and event.
+    let dsc = serde_json::from_str::<DynamicSamplingContext>((*dsc).as_str());
+    let event = serde_json::from_str::<EphemeralEvent>((*event).as_str());
```

Comment:
Instead of parsing some synthetic `EphemeralEvent`, you should just load an actual `Annotated<Event>` directly:

```suggestion
    let event = Annotated::<Event>::from_json((*event).as_str())?;
    let event = event.value().ok_or(<some error>)?;
```